---
title: "Interactive graph-based visualization of genome architecture comparisons"
date: 2019-3-2
tags: [visualization, genome, graphs, graph theory, javafx, java]
header:
  image: "/assets/images/DAGGER/dagger_icon.png"
excerpt: "DAGGER, Brouns lab"
---

## Abstract

The variability between genome architectures, the collection of non-random arrangements of functional elements within a genome, can be used to determine the similarities and differences in a genome comparison to study genome evolution. Long-read sequencing brings the benefit of a more accurate assembly in regions of the genome with repetitive sequences and complex structural variation. Long-reads can be more easily aligned to each other which leads to more accurate comparisons of genome architectures. Instead of aligning sequences, a single graph-based data structure can offer more information of similarities and differences in genome architecture in variable regions. Ptolemy is a tool that is able to compare genome architectures without the need of a reference in a gene graph with vertices representing genes. The interpretation of these gene graphs can however be difficult without visualization, and currently few tools provide biological depth when representing these graphs. In this study we present Directed Acyclic Graphs for GEnomic readability (DAGGER), a tool with a graphical user interface for the interactive visualization of gene graphs from GFA1-formatted files. We show the features of DAGGER by visualizing the genome architecture comparison of three novel Escherichia phages (most closely related to vB_EcoM_AYO145A) in a single gene graph using Ptolemy. We also present the stand-alone tool POET (Ptolemy Output Enhancement Tool). This tool can be used for the further analysis of genome graph data generated by Ptolemy. With help of Ptolemy and POET we partially analyse the genome architecture of a dataset of long-sequencing reads of a chimeric generated bacteriophage of E. coli bacteria BL21phi10. Finally, we present a discussion about the advantages and disadvantages of different layout algorithms to visualize gene graphs.


## Contents

- Abstract
- List of abbreviations

1. Introduction
   1. The benefits of long-read sequencing for comparative genomics
   2. The use of gene graphs for comparative genomics
   3. Visualizing gene graphs
   4. Project aim
2. Methods
   1. Development of DAGGER
      1. Sugiyama Layout
      2. Circular Layout
   2. Development of POET
   3. Three novel coliphage dataset
   4. BL21phi10 chimeric phage dataset
3. Results
   1. DAGGER: Visualization of graph-based genome architecture comparisons
   2. POET: Long-read gene graph analysis
   3. Analysis of the three novel coliphage dataset
   4. Analysis of the BL21phi10 chimeric phage datase
4. Discussion
5. Conclusion

- Acknowledgements
- References


## List of abbreviations
BANDAGE	a Bioinformatics Application for Navigating De novo Assembly Graphs Easily

DAGGER	Directed Acyclic Graphs for GEnomic Readability

GFA1		Graphical Fragment Assembly format 1

JUNG		Java Universal Network/Graph framework

POET		Ptolemy Output Enhancement Tool

## 1 Introduction
The first sequencing of a genome (bacteriophage MS2) was reported in 1976 (Fiers et al., 1976). The upcoming of comparative genomics would however be much later, starting with the whole genome sequencing of _Mycoplasma genitalium_ (Fraser et al., 1995) and _Haemophilus influenza_ (Fleischmann et al., 1995). Currently, comparative genomics is widely used when analysing sequenced genomes (Koonin & Galperin, 2013). The field of comparative genomics aims to compare the genetic content of two or more species to identify the similarities and differences between genomes. The totality of non-random arrangements of functional elements in a genome can be described as genome architecture (Koonin, 2009). The study of genome evolution (the process by which a genome changes in structure over time through mutation, horizontal gene transfer and sexual reproduction) compares genome architectures between closely and distantly related genomes. This comparison can indicate the similarities and differences through evolution and lead to findings of how a genome is able to evolve through time.

A sequencing run generates data in the form of reads. A read is an inferred sequence of base pairs of as given molecule. Reads that share an overlap can be assembled into contigs. A contig can be described as a collection of partially or entirely overlapping reads that together represent a consensus sequence of a genetic fragment. The genome assembly can contain multiple contigs, which indicates the presence of a gap. To bridge the gap between these contigs it is needed to create a scaffold (__Figure 1__) (Fierst, 2015; Waterston, Lander, & Sulston, 2002). Outside of these contigs and scaffolds it is also possible to visualize the result of the assembly as a graph. Assembly graphs provide extra layer of information, as it contains connections between reads (Wick, Schultz, Zobel, & Holt, 2015).

<a href="/assets/images/DAGGER/circlelayout.PNG"><img src="/assets/images/DAGGER/circlelayout.PNG"></a>
<a href="/assets/images/DAGGER/sugiyamalayout.PNG"><img src="/assets/images/DAGGER/sugiyamalayout.PNG"></a>